packages/unpublished-eslint-rules/index.js

```diff
@@ -0,0 +1,285 @@
+                  let [start, end] = declaration.range;
+                  removals.push([start, end + 1]);
+
+                  // if this is the import we are removing
+                  // we have to alter where we will do insertion
+                  if (lastImport === declaration) {
+                    if (declarations.length > 1) {
+                      lastImport = declarations[declarations.length - 2];
+                    } else {
+                      lastImport = null;
+                    }
+                  }
+                } else {
+                  let removedRanges = [];
+                  typeImports.forEach(specifier => {
+                    removedRanges.push(getRangeToRemove(specifier));
+                  });
+                  removals = collapseRanges(removedRanges);
+                }
+
+                context.report({
+                  message,
+                  node: declaration,
+                  fix(fixer) {
+                    let changes = [];
+                    if (lastImport) {
+                      // place at the end of all import statements
+                      changes.push(fixer.insertTextAfter(lastImport, newLines));
+                    } else {
+                      // no more ES imports remain after we remove this one
+                      changes.push(fixer.insertTextBeforeRange(declaration.range, newLines));
+                    }
+                    removals.forEach(range => {
+                      changes.push(fixer.removeRange(range));
+                    });
+                    return changes;
+                  },
+                });
+              }
+            }
+            // console.log('\nProgram:exitEnd\n======\n\n\n');
+          },
+          ImportDeclaration(node) {
+            declarations.push(node);
+            lastImport = node;
+          },
+          ImportDefaultSpecifier(node) {
+            imports[node.local.name] = {
+              local: node.local,
+              imported: node.imported || node.local,
+              from: node.parent.source,
+              isDefault: true,
+            };
+          },
+          ImportSpecifier(node) {
+            imports[node.local.name] = {
+              local: node.local,
+              imported: node.imported || node.local,
+              from: node.parent.source,
+              isDefault: false,
+            };
+          },
+          Identifier(node) {
+            let type = node.parent.type;
+
+            if (type === 'ImportSpecifier' || type === 'ImportDefaultSpecifier') {
+              // console.log(`found ${node.name} for import type ${type}`);
+            } else if (type === 'TSTypeReference') {
+              // console.log(`found type reference ${node.name}`);
+              // do nothing, this Identifier is handled elsewhere
+            } else if (imports[node.name] !== undefined) {
+              const scope = context.getScope();
+              if (scope.type === 'class') {
```

Comment:
Lets make the top if a switch or something like that, to make these nested ifs easier to read